### ISSUE TITLE
Build errors in Scheduler (Linux/ gcc compatibility issue)

### ISSUE THREAD

**Issue by stoomart** [Zilliqa/Zilliqa#21]
Fails to build from the tar file Zilliqa-Durianv0.1.tar.gz on a fresh install of Fedora 27 x86_64, see attached log.

Installed packages:
```
boost-devel.x86_64                    1.64.0-4.fc27
cmake.x86_64                          3.10.1-4.fc27
gcc-c++.x86_64                        7.2.1-2.fc27
jsoncpp-devel.x86_64                  1.8.3-1.fc27
leveldb-devel.x86_64                  1.18-1.fc26
openssl-devel.x86_64                  1:1.1.0g-1.fc27
```

[Durian_build.log](https://github.com/Zilliqa/Zilliqa/files/1611276/Durian_build.log)

**Comment by sidutta** [comment]
Can you try checking whether explicitly including functional header helps and also that you don't have any other C++11 related compiler issue?

**Comment by stoomart** [comment]
Sorry I'm not a developer, I'm not familiar with how to explicitly include functional headers. I ran a C++11 compiler verification script and all the checks came back OK.

I have since tried building on CentOS 7.4.1708, which also failed complaining about c++14, which I determined was due to gcc-c++ 4.8.5 (4.9 is required).

The build got much further on Fedora 25 (uses gcc-c++ 6.4.1 instead of 7.2.1 in F27), but ends up complaining about undefined references to LevelDB::..., which seem to be defined in the LevedDB.h file. I will keep working on this, let me know if you have any quick pointers.

**Comment by stoomart** [comment]
Experienced the same issue originally reported on a fresh install of Ubuntu 17.10, which uses gcc 7.2. Installing Ubuntu 17.04 now, which uses gcc 6.3.

[Durian_build_ubuntu-1710.log](https://github.com/Zilliqa/Zilliqa/files/1613803/Durian_build_ubuntu-1710.log)

**Comment by stoomart** [comment]
Built and executed test scripts successfully on Ubuntu 17.04, looks like there's a specific library version/path dependency somewhere.

**Comment by sidutta** [comment]
Thanks for sharing your findings. I use gcc 5.4. We could use this thread to discuss compatibility issues for most major versions of gcc and linux. I have edited the title to reflect this.

**Comment by MastaG** [comment]
Yes, adding #include <functional> to src/libUtils/Scheduler.h fixed this issue.
However lt's not linking correctly with leveldb, while both leveldb and leveldb-devel are installed.

**[ 72%] Linking CXX executable Test_Transaction
../../src/depends/libDatabase/libOverlay.so: undefined reference to `LevelDB::Lookup[abi:cxx11](dev::FixedHash<32u> const&) const'
../../src/depends/libDatabase/libOverlay.so: undefined reference to `LevelDB::Exists(dev::FixedHash<32u> const&) const'
../../src/depends/libDatabase/libOverlay.so: undefined reference to `LevelDB::Lookup[abi:cxx11](vector_ref<unsigned char const> const&) const'
../../src/depends/libDatabase/libOverlay.so: undefined reference to `LevelDB::BatchInsert(std::unordered_map<dev::FixedHash<32u>, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int>, std::hash<dev::FixedHash<32u> >, std::equal_to<dev::FixedHash<32u> >, std::allocator<std::pair<dev::FixedHash<32u> const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> > > >&, std::unordered_map<dev::FixedHash<32u>, std::pair<std::vector<unsigned char, std::allocator<unsigned char> >, bool>, std::hash<dev::FixedHash<32u> >, std::equal_to<dev::FixedHash<32u> >, std::allocator<std::pair<dev::FixedHash<32u> const, std::pair<std::vector<unsigned char, std::allocator<unsigned char> >, bool> > > >&)'
collect2: fout: ld gaf exit-status 1 terug
make[2]: *** [tests/Data/CMakeFiles/Test_Transaction.dir/build.make:105: tests/Data/Test_Transaction] Fout 1
make[1]: *** [CMakeFiles/Makefile2:1777: tests/Data/CMakeFiles/Test_Transaction.dir/all] Fout 2
make: *** [Makefile:84: all] Fout 2**

**Comment by sidutta** [comment]
@MastaG May be try explicitly adding -L <path> to the root CMakeLists.txt. 

if(NOT IS_LOOKUP_NODE)
    if(${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
        SET( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${GCC_COVERAGE_COMPILE_FLAGS} -pthread -std=c++14 -Wall -DSTAT_TEST -ggdb" )
        SET( CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${GCC_COVERAGE_LINK_FLAGS} -pthread -ljsoncpp -lboost_system -lboost_filesystem -std=c++14 -Wall -DSTAT_TEST -L path_here" )

A proper solution may be something as in https://github.com/Zilliqa/Zilliqa/blob/macosx-support/CMakeLists.txt

**Comment by MastaG** [comment]
Doesn't seem to work. I've tried to add -lleveldb and -L/usr/lib64 to all four linker lines in the root CMakeLists.txt.
I can confirm that the libleveldb.so is present in /usr/lib64 and so are the header files in /usr/include.

Still the same error.

**Comment by MastaG** [comment]
Should I open new issue for the leveldb problem?

**Comment by MastaG** [comment]
Anyways, here the exact error:

[ 71%] Linking CXX executable Test_Transaction
cd /home/mastag/src/Zilliqa/build/tests/Data && /usr/bin/cmake -E cmake_link_script CMakeFiles/Test_Transaction.dir/link.txt --verbose=1
/usr/bin/c++    -pthread -std=c++14 -Wall -DSTAT_TEST -ggdb -g    -pthread -ljsoncpp -lboost_system -lboost_filesystem -std=c++14 -Wall -DSTAT_TEST -rdynamic CMakeFiles/Test_Transaction.dir/Test_Transaction.cpp.o  -o Test_Transaction -Wl,-rpath,/home/mastag/src/Zilliqa/build/src/depends/libTrie:/home/mastag/src/Zilliqa/build/src/depends/common:/home/mastag/src/Zilliqa/build/src/depends/libDatabase ../../src/libData/AccountData/libAccountData.a ../../src/libUtils/libUtils.a ../../src/libData/BlockData/Block/libBlock.a ../../src/libData/BlockData/BlockHeader/libBlockHeader.a ../../src/libCrypto/libCrypto.a -lcrypto ../../src/depends/libTrie/libTrie.so ../../src/depends/libDatabase/libDatabase.a ../../src/libUtils/libUtils.a ../../src/depends/common/libCommon.so -lboost_filesystem -lboost_system ../../src/depends/libDatabase/libOverlay.so -L/usr/lib64 -lleveldb 
../../src/depends/libDatabase/libOverlay.so: undefined reference to `LevelDB::Lookup[abi:cxx11](dev::FixedHash<32u> const&) const'
../../src/depends/libDatabase/libOverlay.so: undefined reference to `LevelDB::Exists(dev::FixedHash<32u> const&) const'
../../src/depends/libDatabase/libOverlay.so: undefined reference to `LevelDB::Lookup[abi:cxx11](vector_ref<unsigned char const> const&) const'
../../src/depends/libDatabase/libOverlay.so: undefined reference to `LevelDB::BatchInsert(std::unordered_map<dev::FixedHash<32u>, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int>, std::hash<dev::FixedHash<32u> >, std::equal_to<dev::FixedHash<32u> >, std::allocator<std::pair<dev::FixedHash<32u> const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> > > >&, std::unordered_map<dev::FixedHash<32u>, std::pair<std::vector<unsigned char, std::allocator<unsigned char> >, bool>, std::hash<dev::FixedHash<32u> >, std::equal_to<dev::FixedHash<32u> >, std::allocator<std::pair<dev::FixedHash<32u> const, std::pair<std::vector<unsigned char, std::allocator<unsigned char> >, bool> > > >&)'
collect2: fout: ld gaf exit-status 1 terug

**Comment by stoomart** [comment]
The fix #47 resolved the original issue when building from a fresh git clone on Fedora 27, Ubuntu 17.10, and CentOS 7.

I was able to successfully bypass the LevelDB build errors in the 4 failing tests (Test_Account, Test_Transaction, Test_LookupNodeForDSBlock, and Test_LookupNodeForTxBlock) by commenting out their directories in test/CMakeLists.txt:
```
#add_subdirectory (Data)
#add_subdirectory (Lookup)
```

I was able to build and run locally on CentOS by installing gcc 5 and building in a development terminal:
```
sudo yum install centos-release-scl epel-release
sudo yum install boost-devel devtoolset-4-gcc* git jsoncpp-devel cmake leveldb-devel openssl-devel psmisc
scl enable devtoolset-4 bash
./build.sh
```

Now I'm just working on some freezes that appear to be related to user max process limits/nproc.
